### PR TITLE
Add support for Go 1.27 & jsonv2

### PR DESCRIPTION
**Pull request by anderseknert**
Makes OPA build and pass its tests on Go 1.27, while keeping Go 1.25 and 1.26 working. JSON output is unchanged on every supported version.

Go 1.27 json package honours `encoding.TextAppender`. Many v1 ast types implement AppendText to build their Rego string cheaply, so on 1.27 they would have marshalled as Rego text. Files built only with 1.27 now implement MarshalJSONTo.

Library users should keep using `json.Marshal` etc. The MarshalJSONTo methods are implementation details, are absent from 1.25 and 1.26 builds, and may change.

### CODE REVIEW

File: v1/util/jsonv2.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+// JsonEqual returns true if the canonical JSON encoding of a and b are equal,
+// meaning that a and b are compared without regard to things like whitespace,
+// key order, etc. For more details, see [jsontext.Value.Canonicalize].
+func JsonEqual[A, B stringOrBytes](a A, b B) bool {
+	v1, v2 := jsontext.Value(a), jsontext.Value(b)
+
+	v1.Canonicalize()
+	v2.Canonicalize()
```
Comment (johanfylling) [inline v1/util/jsonv2.go]: nice 👍

**Comment by charlieegan3** [comment]
I've added a few more edits here to get the tests passing on 1.26 and 1.27. Johan is also going to do a review so we can get this in for the release 🤞

File: cmd/parse_test.go [hunk 2]
```diff
@@ -241,21 +238,8 @@ p = 1
...
-	gotLines := strings.Split(string(stdout), "\n")
-	wantLines := strings.Split(expectedOutput, "\n")
-	min := len(gotLines)
-	if len(wantLines) < min {
-		min = len(wantLines)
-	}
-
-	for i := range min {
-		if gotLines[i] != wantLines[i] {
-			t.Fatalf("Expected line %d to be\n%v\n, got\n%v", i, wantLines[i], gotLines[i])
```
Comment (johanfylling) [inline cmd/parse_test.go:253]: Looks like we might be losing some precision here; which might add some time when troubleshooting. OTOH, it's fairly easy to just diff the strings yourself; but you might not be able to see the root cause directly in the test output.

File: cmd/parse_test.go [hunk 3]
```diff
@@ -625,15 +609,14 @@ func TestParseOutputWithNotImport(t *testing.T) {
 				t.Fatalf("Expected no stderr output, got:\n%s\n", string(stderr))
 			}
 
-			if diff := cmp.Diff(tc.exp, string(stdout)); diff != "" {
-				t.Errorf("unexpected result (-want, +got):\n%s", diff)
+			if !util.JsonEqual(stdout, tc.exp) {
```
Comment (johanfylling) [inline cmd/parse_test.go]: Since this test passes, I assume the `formats.Pretty` format will just pass through the `jsontext` `Value.Canonicalize()` unchanged.

An issue I can see here though is that even though it's convenient to do a canonicalized comparison, it also means that we won't detect formatting regression. E.g. we start dropping newlines or whitespaces. It won't break any software consumers, as the json is still valid, but it might affect readability.

File: internal/semver/semver.go [hunk 4]
```diff
@@ -104,7 +104,7 @@ func Compare(a, b string) int {
 // AppendText appends the textual representation of the version to b and returns the extended buffer.
```
Comment (johanfylling) [inline internal/semver/semver.go]: AppendString

File: internal/semver/semver_test.go [hunk 5]
```diff
@@ -165,7 +165,7 @@ func BenchmarkAppendText(b *testing.B) {
 	v := MustParse("1.2.3-alpha.1+build.123")
 
 	for b.Loop() {
-		_, err := v.AppendText(nil)
+		_, err := v.AppendString(nil)
```
Comment (johanfylling) [inline internal/semver/semver_test.go:168]: Test should be renamed to `BenchmarkAppendString`

File: internal/semver/semver_test.go [hunk 6]
```diff
@@ -181,7 +181,7 @@ func BenchmarkAppendTextPreAllocated(b *testing.B) {
 	buf := make([]byte, 0, 32)
 
 	for b.Loop() {
-		if buf, err = v.AppendText(buf); err != nil {
+		if buf, err = v.AppendString(buf); err != nil {
```
Comment (johanfylling) [inline internal/semver/semver_test.go:184]: `BenchmarkAppendStringPreAllocated`

File: v1/ast/location/location_jsonv2.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+				break
+			}
+			switch key {
+			case "file":
+				loc.File = tok.String()
+			case "row":
+				loc.Row, err = readInt(tok, "row")
+			case "col":
+				loc.Col, err = readInt(tok, "col")
+			case "text", "tabs": // marked as ignored
```
Comment (johanfylling) [inline v1/ast/location/location_jsonv2.go]: No explicit skip for `offset`? `text` is here because of the `jsonOptions.IncludeLocationText` condition in `MarshalJSONTo`; why is `tabs` here?

File: v1/ast/location/location_jsonv2.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+	if jsonOptions.IncludeLocationText {
+		e.WriteToken(jsontext.String("text"))
+		e.WriteToken(jsontext.String(base64.StdEncoding.EncodeToString(loc.Text)))
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func UnmarshalFromFn(d *jsontext.Decoder, loc *Location) error {
```
Comment (johanfylling) [inline v1/ast/location/location_jsonv2.go]: The `json:` tags on `Location` aren't enough for unmarshalling?

Comment (johanfylling) [reply v1/ast/location/location_jsonv2.go]: Oh, this isn't [UnmarshalJSONFrom](https://pkg.go.dev/encoding/json/v2#UnmarshalerFrom). So this won't be picked up through introspection? What's the purpose of this? External tool use? Are the explicit `text`/`tabs` ignores part of the reason?

File: v1/ast/annotations.go [hunk 9]
```diff
@@ -192,64 +191,6 @@ func (a *Annotations) GetTargetPath() Ref {
...
-
-	if len(a.Custom) > 0 {
-		data["custom"] = a.Custom
-	}
-
-	if len(a.Labels) > 0 {
-		data["labels"] = a.Labels
-	}
-
-	if astJSON.GetOptions().MarshalOptions.IncludeLocation.Annotations {
```
Comment (johanfylling) [inline v1/ast/annotations.go:244]: We're dealing with this somewhere else now?

Comment (johanfylling) [reply v1/ast/annotations.go:244]: Ah, ok, moved to `policy_json.go`/`policy_jsonv2.go`; I would have expected `annotations_json.go` etc.

File: v1/ast/policy_jsonv2.go [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+
+	if a.Location != nil && astJSON.GetOptions().MarshalOptions.IncludeLocation.Annotations {
+		e.WriteToken(jsontext.String("location"))
+		a.Location.MarshalJSONTo(e)
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func (a *Annotations) MarshalJSON() ([]byte, error) {
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: We have both `MarshalJSONTo` and this to not break the exported API got Go 1.27?

File: v1/ast/policy_jsonv2.go [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+	"github.com/open-policy-agent/opa/v1/util"
+)
+
+func (a *Annotations) MarshalJSONTo(e *jsontext.Encoder) error {
+	e.WriteToken(jsontext.BeginObject)
+
+	if a == nil {
+		e.WriteToken(jsontext.String("scope"))
+		e.WriteToken(jsontext.String(""))
+		return e.WriteToken(jsontext.EndObject)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: [nit] I suppose this could be a single deferred call, but we don't have that many exits to contend with anyways 🤷

File: v1/ast/policy_jsonv2.go [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+		json.MarshalEncode(e, a.Custom)
+	}
+
+	if len(a.Labels) > 0 {
+		e.WriteToken(jsontext.String("labels"))
+		json.MarshalEncode(e, a.Labels)
+	}
+
+	e.WriteToken(jsontext.String("scope"))
+	e.WriteToken(jsontext.String(a.Scope))
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: [nit] stylistically, It'd be nice to put non-optional fields either first or last, and not mix-and-match.

File: v1/ast/policy_jsonv2.go [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+		e.WriteToken(jsontext.String("title"))
+		e.WriteToken(jsontext.String(a.Title))
+	}
+
+	if a.Location != nil && astJSON.GetOptions().MarshalOptions.IncludeLocation.Annotations {
+		e.WriteToken(jsontext.String("location"))
+		a.Location.MarshalJSONTo(e)
+	}
+
+	return e.WriteToken(jsontext.EndObject)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: Aren't we missing the `compile` field?

File: v1/ast/policy_jsonv2.go [hunk 14]
```diff
@@ -0,0 +1,586 @@
+//go:build go1.27
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go:1]: No `AnnotationsRef` marshaler?

File: v1/ast/policy_jsonv2.go [hunk 15]
```diff
@@ -0,0 +1,586 @@
...
+	}
+
+	if astJSON.GetOptions().MarshalOptions.IncludeLocation.Every {
+		e.WriteToken(jsontext.String("location"))
+		if q.Location != nil {
+			if err := q.Location.MarshalJSONTo(e); err != nil {
+				return err
+			}
+		} else {
+			e.WriteToken(jsontext.Null)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: Why do we explicitly write `null` here, and not for other `location`s? E.g. `Rule`, `Head`, `Import`, etc.

File: v1/ast/policy_jsonv2.go [hunk 16]
```diff
@@ -0,0 +1,586 @@
...
+	}
+
+	return enc.WriteToken(jsontext.EndObject)
+}
+
+func (a *LogicalAnd) MarshalJSONTo(e *jsontext.Encoder) error {
+	e.WriteToken(jsontext.BeginObject)
+	e.WriteToken(jsontext.String("type"))
+	e.WriteToken(jsontext.String("and"))
+	e.WriteToken(jsontext.String("lhs"))
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: [nit] consider separating fields with a newline for readability.

File: v1/ast/policy_jsonv2.go [hunk 17]
```diff
@@ -0,0 +1,586 @@
...
+		e.WriteToken(jsontext.String("location"))
+		if err := a.Location.MarshalJSONTo(e); err != nil {
+			return err
+		}
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func (a *LogicalAnd) UnmarshalJSON(bs []byte) error {
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go:359]: So we don't need to pair a `MarshalJSONTo` to a `UnmarshalJSONFrom`?

File: v1/ast/term_jsonv2.go [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+// MarshalJSON returns JSON encoded bytes representing s.
+func (s *set) MarshalJSON() ([]byte, error) {
+	return util.MarshalMarshalerTo(s)
+}
+
+func (o *LogicalOr) MarshalJSON() ([]byte, error) {
+	return util.MarshalMarshalerTo(o)
+}
+
+func (o *LogicalOr) MarshalJSONTo(e *jsontext.Encoder) error {
```
Comment (johanfylling) [inline v1/ast/term_jsonv2.go]: How come this is located here, but `LogicalAnd.MarshalJSONTo()` is in `policy_json2.go`?

File: v1/ast/policy_jsonv2.go [hunk 19]
```diff
@@ -0,0 +1,586 @@
...
+	if len(s.Path) > 0 {
+		e.WriteToken(jsontext.String("path"))
+		e.WriteToken(jsontext.BeginArray)
+		for _, t := range s.Path {
+			// No location here, as it's incorrect
+			e.WriteToken(jsontext.BeginObject)
+			e.WriteToken(jsontext.String("type"))
+			e.WriteToken(jsontext.String(ValueName(t.Value)))
+			e.WriteToken(jsontext.String("value"))
+			marshalValueTo(e, t.Value)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: Willfully ignoring the returned error?

File: v1/ast/policy_jsonv2.go [hunk 20]
```diff
@@ -0,0 +1,586 @@
...
+		return err
+	}
+	return unmarshalLogical("and", &a.Lhs, &a.Rhs, &a.ExplicitLhs, &a.ExplicitRhs, v)
+}
+
+func (w *With) MarshalJSONTo(e *jsontext.Encoder) error {
+	e.WriteToken(jsontext.BeginObject)
+
+	e.WriteToken(jsontext.String("target"))
+	w.Target.MarshalJSONTo(e)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: We don't want to return the error here? This seems to be a trend. Maybe it's mostly inconsequential, but for `Term.MarshalJSONTo()` there seems to be a custom error.

File: v1/ast/policy_jsonv2.go [hunk 21]
```diff
@@ -0,0 +1,586 @@
...
+	if len(annotations) > 0 {
+		// TODO: filter in place, or range over function
+		var packageAnnotations []*Annotations
+		for _, a := range annotations {
+			if a.Scope == "package" || a.Scope == "subpackages" {
+				packageAnnotations = append(packageAnnotations, a)
+			}
+		}
+		if len(packageAnnotations) > 0 {
+			e.WriteToken(jsontext.String("annotations"))
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: This is a new field not present in the old package marshalling?

File: v1/ast/policy_jsonv2.go [hunk 22]
```diff
@@ -0,0 +1,586 @@
...
+	if r.Else != nil {
+		e.WriteToken(jsontext.String("else"))
+		r.Else.MarshalJSONTo(e)
+	}
+
+	e.WriteToken(jsontext.String("head"))
+	r.Head.MarshalJSONTo(e)
+
+	e.WriteToken(jsontext.String("body"))
+	r.Body.MarshalJSONTo(e)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: [nit] grouping required and optional fields would be nice.

File: v1/ast/policy_jsonv2.go [hunk 23]
```diff
@@ -0,0 +1,586 @@
...
+
+	if astJSON.GetOptions().MarshalOptions.IncludeLocation.Head && h.Location != nil {
+		e.WriteToken(jsontext.String("location"))
+		h.Location.MarshalJSONTo(e)
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func (c Call) MarshalJSONTo(e *jsontext.Encoder) (err error) {
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go:196]: Are we missing `MarshalJSON` for `Call`?

File: v1/ast/marshal_jsonv2_test.go [hunk 24]
```diff
@@ -0,0 +1,1297 @@
+// Copyright 2026 The OPA Authors.  All rights reserved.
```
Comment (johanfylling) [inline v1/ast/marshal_jsonv2_test.go:1]: Looks like we're leaving out `TestAnnotations_MarshalJSON_Compile`. Which I suppose matches the omitted `compile` field in annotation marshalling. So that _is_ a conscious choice then?

File: v1/ast/policy_appenders.go [hunk 25]
```diff
@@ -22,7 +22,9 @@ func (m *Module) AppendText(buf []byte) ([]byte, error) {
 		if annotations.Scope == "package" || annotations.Scope == "subpackages" {
 			buf = append(buf, "# METADATA\n# "...)
 			buf = append(buf, annotations.String()...)
-			buf = append(buf, '\n')
+			if buf[len(buf)-1] != '\n' {
+				buf = append(buf, '\n')
+			}
```
Comment (johanfylling) [inline v1/ast/policy_appenders.go]: Drive-by bug fix?

File: v1/ast/policy_logical_jsonv2_test.go [hunk 26]
```diff
@@ -0,0 +1,720 @@
...
+}
+
+func newLogicalOr(lhs, rhs string) *LogicalOr {
+	return &LogicalOr{
+		Lhs: NewBody(NewExpr(VarTerm(lhs))),
+		Rhs: NewBody(NewExpr(VarTerm(rhs))),
+	}
+}
+
+func TestLogicalAnd_Compare(t *testing.T) {
```
Comment (johanfylling) [inline v1/ast/policy_logical_jsonv2_test.go]: Not all tests in here are related to marshalling; maybe we shouldn't duplicate what we don't have to?

File: v1/ast/policy_logical_jsonv2_test.go [hunk 27]
```diff
@@ -0,0 +1,720 @@
...
+
+func TestLogical_Or_Hash_StableForEqualOperands(t *testing.T) {
+	c := newLogicalOr("x", "y")
+	d := newLogicalOr("x", "y")
+	if c.Hash() != d.Hash() {
+		t.Fatalf("equal LogicalOr should hash identically, got %d vs %d", c.Hash(), d.Hash())
+	}
+}
+
+func TestLogical_String(t *testing.T) {
```
Comment (johanfylling) [inline v1/ast/policy_logical_jsonv2_test.go]: Looks like a bunch of test cases have been dropped in here when compared to those in `policy_logical_test.go`.

File: v1/ast/term_json.go [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+	if astJSON.GetOptions().MarshalOptions.IncludeLocation.Not {
+		if n.Location != nil {
+			data["location"] = n.Location
+		}
+	}
+
+	return json.Marshal(data)
+}
+
+func (n *Not) UnmarshalJSON(bs []byte) error {
```
Comment (johanfylling) [inline v1/ast/term_json.go:88]: This is only available <go1.27?

**Review by johanfylling — COMMENTED** [review]
Lots of good stuff in here!

Some thoughts.

File: v1/ast/marshal_jsonv2_test.go
Comment (johanfylling) [inline v1/ast/marshal_jsonv2_test.go:1]: Shouldn't this be `marshal_jsonv2_test.go`?

File: v1/ast/policy_jsonv2.go [hunk 29]
```diff
@@ -0,0 +1,586 @@
...
+		}
+		if len(packageAnnotations) > 0 {
+			e.WriteToken(jsontext.String("annotations"))
+			util.WriteMarshalerToArray(e, packageAnnotations)
+		}
+	}
+
+	if astJSON.GetOptions().MarshalOptions.IncludeLocation.Package {
+		e.WriteToken(jsontext.String("location"))
+		pkg.Location.MarshalJSONTo(e)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: We don't wrap this in a `nil` check. Could we blow up here?

File: v1/ast/policy_jsonv2.go [hunk 30]
```diff
@@ -0,0 +1,586 @@
...
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func (m *Module) MarshalJSONTo(e *jsontext.Encoder) error {
+	e.WriteToken(jsontext.BeginObject)
+
+	e.WriteToken(jsontext.String("package"))
+	m.Package.marshalJSONTo(e, m.Annotations)
```
Comment (johanfylling) [inline v1/ast/policy_jsonv2.go]: Shouldn't this be writing the package, not annotations?

File: v1/ast/policy_json.go [hunk 31]
```diff
@@ -0,0 +1,366 @@
...
+	Body        Body           `json:"body"`
+	Default     bool           `json:"default,omitempty"`
+	Else        *Rule          `json:"else,omitempty"`
+	Head        *Head          `json:"head"`
+	Location    *Location      `json:"location,omitempty"`
+}
+
+// UnmarshalJSON parses bs and stores the result in mod. The rules in the module
+// will have their module pointer set to mod.
+func (mod *Module) UnmarshalJSON(bs []byte) error {
```
Comment (johanfylling) [inline v1/ast/policy_json.go:44]: Shouldn't there be a corresponding  function for Go1.27?

File: v1/ast/policy_json.go [hunk 32]
```diff
@@ -0,0 +1,366 @@
...
+	}
+
+	if len(rr.Description) > 0 {
+		d["description"] = rr.Description
+	}
+
+	return json.Marshal(d)
+}
+
+func (ar *AnnotationsRef) MarshalJSON() ([]byte, error) {
```
Comment (johanfylling) [inline v1/ast/policy_json.go]: Don't see this for Go1.27.

**Review by johanfylling — COMMENTED** [review]
Some additional notes

File: v1/ast/marshal_jsonv2_test.go
Comment (johanfylling) [reply v1/ast/marshal_jsonv2_test.go:1]: This might explain why some of the issues found aren't exposed as failing tests 🤔

File: v1/ast/policy_appenders.go (hunk 25 above)
Comment (charlieegan3) [reply v1/ast/policy_appenders.go]: I dug into this a little, this is related to the new functionality for values vs documents having trailing new lines.

I think we can use json.Marshal directly, as in https://github.com/open-policy-agent/opa/pull/8947/commits/b9fc0c28cc7366f5c6f53cb35533491afc3306e3 to avoid much of this. It seems to be fewer allocations too.

File: v1/ast/policy_json.go (hunk 32 above)
Comment (charlieegan3) [reply v1/ast/policy_json.go]: Added in e5eb59684

File: v1/ast/marshal_jsonv2_test.go
Comment (charlieegan3) [reply v1/ast/marshal_jsonv2_test.go:1]: Yeah, ee7e2921b moves the test to the correct location, following commits address some issues

File: v1/ast/policy_jsonv2.go (hunk 14 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go:1]: added in e5eb59684

File: v1/ast/policy_jsonv2.go (hunk 13 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: Fixed in c62879d03, test added.

File: v1/ast/marshal_jsonv2_test.go (hunk 24 above)
Comment (charlieegan3) [reply v1/ast/marshal_jsonv2_test.go:1]: test added in c62879d03, fixes missing field

File: v1/ast/policy_jsonv2.go (hunk 29 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: Nil guards added in 9a48c4208

File: v1/ast/policy_jsonv2.go (hunk 21 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: Fixed in 591b4839c, this should not have been added I think

File: v1/ast/policy_jsonv2.go (hunk 30 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: related to 591b4839c where annotations is removed from package.

File: v1/ast/policy_json.go (hunk 31 above)
Comment (charlieegan3) [reply v1/ast/policy_json.go:44]: a592fe8b8 addresses this issue and inconsistencies in the include location settings. there is minor difference in 1.26 and 1.27 here. see commit

File: v1/ast/term_json.go (hunk 28 above)
Comment (charlieegan3) [reply v1/ast/term_json.go:88]: it was missing , added in a592fe8b827a1e821fdec2d97d4b9398646aa0dd

File: v1/ast/policy_logical_jsonv2_test.go (hunk 27 above)
Comment (charlieegan3) [reply v1/ast/policy_logical_jsonv2_test.go]: 3ace43c9a fixes this, jsonEqual addresses the issue so no need for split test here.

File: v1/ast/policy_logical_jsonv2_test.go (hunk 26 above)
Comment (charlieegan3) [reply v1/ast/policy_logical_jsonv2_test.go]: 3ace43c9ab2491e9b48211148da77694e224dbb3 removes dupes

File: v1/ast/location/location_jsonv2.go (hunk 8 above)
Comment (charlieegan3) [reply v1/ast/location/location_jsonv2.go]: I wasn't able to see why this was needed, so have removed it in 0fb2049c5401ab7cd9c44bc6c4ec06c4acac8ded.

File: internal/semver/semver.go (hunk 4 above)
Comment (charlieegan3) [reply internal/semver/semver.go]: Fixed in c5e5a0428

File: internal/semver/semver_test.go (hunk 5 above)
Comment (charlieegan3) [reply internal/semver/semver_test.go:168]: Fixed in 707c95dbc

File: internal/semver/semver_test.go (hunk 6 above)
Comment (charlieegan3) [reply internal/semver/semver_test.go:184]: Fixed in 707c95dbc

File: v1/ast/annotations.go (hunk 9 above)
Comment (charlieegan3) [reply v1/ast/annotations.go:244]: Moved to own file in 81046feb9

File: v1/ast/term_jsonv2.go (hunk 18 above)
Comment (charlieegan3) [reply v1/ast/term_jsonv2.go]: Moved in 81046feb9 to own annotations file.

File: v1/ast/policy_jsonv2.go (hunk 19 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: I have made a few adjustments to how all this works in d5e8ee89a, please take a look, will leave this one open.

File: v1/ast/policy_jsonv2.go (hunk 20 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: this is also related to changes in d5e8ee89a but marking as resolved as another thread on topic is still open

File: v1/ast/policy_jsonv2.go (hunk 15 above)
Comment (charlieegan3) [reply v1/ast/policy_jsonv2.go]: related to 9a48c4208

File: v1/util/jsonv2.go (hunk 1 above)
Comment (charlieegan3) [reply v1/util/jsonv2.go]: This is moved to test helper in 9380f31f1

File: cmd/parse_test.go (hunk 3 above)
Comment (charlieegan3) [reply cmd/parse_test.go]: wdyt about af94143fb, a small test to check the 1.27 format with a todo to come back once it's the default, and move the other parse_test into there to avoid a massive duplication.

File: cmd/parse_test.go (hunk 2 above)
Comment (charlieegan3) [reply cmd/parse_test.go:253]: New test in af94143fb9037d20d1f2f846b609ee04e34871d0 uses cmp.Diff instead.

File: v1/ast/location/location_jsonv2.go (hunk 8 above)
Comment (anderseknert) [reply v1/ast/location/location_jsonv2.go]: One of the benefits of JSON v2 is that marshaling/unmarshaling implementations don’t need to be coupled to the objects they marshal/unmarshal. Ideally (IMHO, obviously) we wouldn’t implement any of the To/From methods on our structs directly, but provide standalone functions to do this without extending the API of existing objects. This is more complicated though, and I only made a start of an attempt in that direction. I think it’s 100% OK implement this the easier way for the next release, but it’d be good to note that external clients must not rely on the To/From methods directly, as we might not want to keep those around later.

**Comment by charlieegan3** [comment]
x-ref https://github.com/open-policy-agent/regal/pull/2066, related to 6cf1d33d38256e9111985c6ba4e161402c7ebc63, a592fe8b8 and historic fix in #7281.

File: cmd/parse_test.go (hunk 2 above)
Comment (johanfylling) [reply cmd/parse_test.go:253]: I think for cmd output tests we might want to use `cmp.Diff` too instead of `util.JsonEqual` like we do here in this file. The reason being that the canonicalization will miss if we start messing up whitespaces and newlines. Not the whole world machine-to-machine, but there is some value to human readability for cmd output.

File: v1/ast/location/location_jsonv2.go [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+package location
+
+import (
+	"encoding/base64"
+	"encoding/json/jsontext"
+
+	astJSON "github.com/open-policy-agent/opa/v1/ast/json"
+)
+
+func (loc *Location) MarshalJSONTo(e *jsontext.Encoder) (err error) {
```
Comment (johanfylling) [inline v1/ast/location/location_jsonv2.go:28]: For other types, we have a `MarshalJSON` that routes into the jsonv2 func. Should we do the same for `Location`?

File: v1/ast/annotations_jsonv2.go [hunk 34]
```diff
@@ -0,0 +1,199 @@
...
+
+	if len(rr.Description) > 0 {
+		e.WriteToken(jsontext.String("description"))
+		e.WriteToken(jsontext.String(rr.Description))
+	}
+
+	return e.WriteToken(jsontext.EndObject)
+}
+
+func (a *AuthorAnnotation) MarshalJSONTo(e *jsontext.Encoder) error {
```
Comment (johanfylling) [inline v1/ast/annotations_jsonv2.go]: Is this strictly necessary? The json tags on the struct aren't enough?

Comment (johanfylling) [inline v1/ast/annotations_jsonv2.go]: This has no `MarshalJSON` counterpart.

File: v1/util/jsonv2.go [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+		return e.WriteToken(jsontext.Null)
+	}
+	return WriteMarshalerToArray(e, items)
+}
+
+// MarshalMarshalerTo provides a MarshalJSON implementation for any type that
+// implements json.MarshalerTo. json.Marshal dispatches to MarshalJSONTo, so this
+// doesn't recurse; the constraint is what guarantees that at compile time.
+func MarshalMarshalerTo[T json.MarshalerTo](v T) ([]byte, error) {
+	return json.Marshal(v)
```
Comment (johanfylling) [inline v1/util/jsonv2.go]: For consistency, could we be passing in [DefaultOptionsV1](https://pkg.go.dev/encoding/json#DefaultOptionsV1) here? Looks like that would keep deterministic map ordering and HTML/JS escaping as-is; or _do_ we want to drop that? 🤔

File: internal/jsonv2/jsonv2.go [hunk 36]
```diff
@@ -0,0 +1,76 @@
...
+		}
+	}
+	return e.WriteToken(jsontext.EndArray)
+}
+
+// WriteField writes the object member name and then v's JSON encoding, so that
+// the member is written and checked in one statement.
+func WriteField[T json.MarshalerTo](e *jsontext.Encoder, name string, v T) error {
+	e.WriteToken(jsontext.String(name))
+	return v.MarshalJSONTo(e)
```
Comment (johanfylling) [inline internal/jsonv2/jsonv2.go:46]: For any `nil` `v`, we'd blow up here, I think; whereas the old marshaling would simply write `null`. Maybe we should add a nill-check up top and also write out `null`?

File: cmd/parse_jsonv2_test.go [hunk 37]
```diff
@@ -0,0 +1,238 @@
...
+//   - rules[0]: 1.26 puts body before head, 1.27 puts head first.
+//   - rules[0].head: 1.26 orders name, value, ref, location; 1.27 orders name,
+//     ref, value, location.
+//
+// This covers one module rather than every parse output shape.
+//
+// TODO: Once only the jsonv2 implementation is used, port the parse_test.go
+// assertions, and those using JsonEqual in v1/ast/policy_test.go and
+// policy_logical_test.go, to byte compares against the 1.27 key order, then
+// delete util.JsonEqual and this file.
```
Comment (johanfylling) [inline cmd/parse_jsonv2_test.go]: `util.JsonEqual` is however a non-test exported function, so it'd be a breaking change to drop it. Maybe we could make it internal instead?

File: v1/ast/term_jsonv2.go [hunk 38]
```diff
@@ -0,0 +1,225 @@
...
+func (s *set) MarshalJSON() ([]byte, error) {
+	return util.MarshalMarshalerTo(s)
+}
+
+func marshalValueTo(e *jsontext.Encoder, val Value) (err error) {
+	switch v := val.(type) {
+	case json.MarshalerTo:
+		err = v.MarshalJSONTo(e)
+	case encoding.TextAppender:
+		buf, _ := v.AppendText(e.AvailableBuffer())
```
Comment (johanfylling) [inline v1/ast/term_jsonv2.go]: Should we be dropping the error here?

Comment (johanfylling) [inline v1/ast/term_jsonv2.go]: Should we be considering quoting here?

File: v1/ast/policy_json.go [hunk 39]
```diff
@@ -0,0 +1,270 @@
...
+import (
+	"encoding/json"
+
+	astJSON "github.com/open-policy-agent/opa/v1/ast/json"
+	"github.com/open-policy-agent/opa/v1/util"
+)
+
+// ruleJSON is used for JSON serialization of Rule to avoid map allocation overhead.
+// Field order is alphabetical to match previous map-based output.
+type ruleJSON struct {
```
Comment (johanfylling) [inline v1/ast/policy_json.go:14]: Should `exprJSON` and `withJSON` also have been moved to the !1.27 branch?

File: v1/ast/annotations_jsonv2.go (hunk 34 above)
Comment (charlieegan3) [reply v1/ast/annotations_jsonv2.go]: This has been removed in f8f23af99

Comment (charlieegan3) [reply v1/ast/annotations_jsonv2.go]: see https://github.com/open-policy-agent/opa/commit/f8f23af9923590b713f13e5e78e6e76c55b8d46f

File: cmd/parse_jsonv2_test.go (hunk 37 above)
Comment (charlieegan3) [reply cmd/parse_jsonv2_test.go]: Fixed in https://github.com/open-policy-agent/opa/pull/8947/commits/f97d32cfb84c0a1edf650f09ebd460d8a302de43

File: v1/util/jsonv2.go (hunk 35 above)
Comment (charlieegan3) [reply v1/util/jsonv2.go]: I guess the downside is that we get things like AllowInvalidUTF8 and AllowDuplicateNames, but then turning these off is simple and not something we need to do as part of this work. For now, I think the goal is looks like v1, works like v2 where possible.

There is actually a bug from the custom annotations field ordering in annotation.String() where obj keys are not the same order that was not caught by an existing test... I think this is only an issue for library consumers rather than in the binary.

I have brought this back in 2d5627e75a079d085ae9e5eada4dc52840a30139.

File: internal/jsonv2/jsonv2.go (hunk 36 above)
Comment (charlieegan3) [reply internal/jsonv2/jsonv2.go:46]: added in 2d5627e75a079d085ae9e5eada4dc52840a30139

File: v1/ast/term_jsonv2.go (hunk 38 above)
Comment (charlieegan3) [reply v1/ast/term_jsonv2.go]: It should be quoting! But we have nothing that hits this since they all have MarshalJSON... 

bit of an edge case, but handled in ed6e7211c38a19591edb4217483f0ef7eff38b97 with a test too.

Comment (charlieegan3) [reply v1/ast/term_jsonv2.go]: fixed in ed6e7211c38a19591edb4217483f0ef7eff38b97

File: v1/ast/policy_json.go (hunk 39 above)
Comment (charlieegan3) [reply v1/ast/policy_json.go:14]: Yup, good spot, fixed in 6a91a982b7c3db56edfc78698d57a0849ac4f24b

File: v1/ast/location/location_jsonv2.go (hunk 33 above)
Comment (charlieegan3) [reply v1/ast/location/location_jsonv2.go:28]: yeah seems they got missed, Number too

3dc8b7297becfd982db982c5eeb0e2d12bcabb09 adds and tests to check for this being set on types with custom marshalling

**Review by johanfylling — APPROVED** [review]
🎉
